src/pip/_internal/build_env.py

```diff
@@ -128,7 +126,13 @@ def __enter__(self):
             'PYTHONPATH': os.pathsep.join(pythonpath),
         })
 
-    def __exit__(self, exc_type, exc_val, exc_tb):
+    def __exit__(
+        self,
+        exc_type,  # type: Optional[Type[BaseException]]
+        exc_val,  # type: Optional[BaseException]
+        exc_tb  # type: Optional[TracebackType]
+    ):
+        # type: (...) -> None
```

Comment:
I think `__exit__` is expected to return `Optional[bool]`, and I'm concerning whether annotation its parameters is worth it since we don't really use any of the `exc*` for anything.

Reply:
> I think `__exit__` is expected to return `Optional[bool]`

I actually took that type annotation from pip itself, https://github.com/pypa/pip/blob/76b865155ee646927e838c377b46643b51ccd2fd/src/pip/_internal/req/req_tracker.py#L82-L88 

but seems like that's wrong, and we return `bool` from it, instead of `Optional[bool]`

Good catch, but I see that in typeshed, the return type is `bool`. 
https://github.com/python/typeshed/blob/e1e99245bb46223928eba68d4fc74962240ba5b4/stdlib/2and3/contextlib.pyi#L75-L77

Could you please point me where you saw the expected return type is `Optional[bool]` ? Perhaps I am not searching correctly.

> I'm concerning whether annotating its parameters is worth it since we don't really use any of the exc* for anything.

But aren't they implicitly used whereever we use the `BuildEnvironment` as a context manager? We can also add `type: ignore` here, but I thought being explicit since we know about the expected types is better.

Reply:
> but seems like that's wrong, and we return `bool` from it, instead of `Optional[bool]`

I recall it from memory without checking, thanks for correcting me.  [Python only requires it to return true-y or false-y values](https://docs.python.org/3/reference/datamodel.html#context-managers), so I agree that as application, `pip` may choose to be as explicit as possible.

Reply:
Thanks, I updated it to return `bool`, and on that note, should the return type of `__enter__` be changed too?

Reply:
On looking further, I see that the vendor libraries using `__exit__`, return both bool and None

https://github.com/pypa/pip/blob/76b865155ee646927e838c377b46643b51ccd2fd/src/pip/_vendor/urllib3/connectionpool.py#L87-L90


https://github.com/pypa/pip/blob/76b865155ee646927e838c377b46643b51ccd2fd/src/pip/_vendor/requests/sessions.py#L420-L421

So maybe `Optional[bool]` might be better here. Honestly, I am not sure myself too.

Reply:
> should the return type of __enter__ be changed too?

Looking at the use case from `wheel_builder` and `distributions.sdist`, I think the current `__enter__` is OK.

About `__exit__`, as showed in Python's doc, I think the interpreter only care about `bool` of the value, so any type having `__bool__` defined is good IMHO.  I am not entirely sure if annotating the return type as `bool` would infer that but there's a low chance that it's the case since it's not described in PEP 484.  Rethinking about it, I believe using `None` as you originally did it might best *document* the context manager here.

Reply:
Agreed, let me revert my commit then.

Reply:
`None` is fine for now -- it's Falsy and that's all we care about TBH; we can revisit cleaning up our context manager annotation in a follow up, if needed.

Reply:
> `None` is fine for now -- it's Falsy and that's all we care about TBH; we can revisit cleaning up our context manager annotations in a follow up, if needed.

In that case, this PR has no more changes needed, and is ready for review, and subsequent approval/merge.